### PR TITLE
Switch 'osc.conf.config' from dict to Options class with type checking

### PR DESCRIPTION
**Pull request by dmach**
Fixes: #1371
Fixes: #1073 
Obsoletes: #1377

### CODE REVIEW

File: osc/conf.py [hunk 1]
```diff
@@ -59,382 +64,1071 @@
...
+
+    @property
+    def data(self):
+        if callable(self._data):
+            # if ``data`` is a function, call it every time the string gets evaluated
+            # we use the password only from time to time to make a session cookie
+            # and there's no need to keep the password in program memory longer than necessary
+            result = self._data()
+            if result is None:
+                raise oscerr.OscIOError(None, "Unable to retrieve password")
```
Comment (dcermak) [inline osc/conf.py:79]: `result` is unused
```suggestion
            result = self._data()
            if result is None:
                raise oscerr.OscIOError(None, "Unable to retrieve password")
            return result
```

File: osc/conf.py [hunk 2]
```diff
@@ -59,382 +64,1071 @@
...
+            """
+            The wrapper to call build tool as root (sudo, su -, ...).
+            If empty, the build tool runs under the current user wich works only with KVM at this moment.
+            """
+        ),
+        ini_key="su-wrapper",
+    )
+
+
+def model_to_rst(cls):
```
Comment (dcermak) [inline osc/conf.py:1039]: `cls` appears to be unused?

File: osc/conf.py [hunk 3]
```diff
@@ -59,382 +64,1071 @@
...
+            return str(int(field.default))
+
+        if origin_type == int:
+            return str(field.default)
+
+        if origin_type == str:
+            return f'"{field.default}"'
+
+        if origin_type == list:
+            return str(field.default)
```
Comment (dcermak) [inline osc/conf.py:1077]:
```suggestion
        if origin_type in (int, list):
            return str(field.default)

        if origin_type == str:
            return f'"{field.default}"'
```

File: osc/util/models.py [hunk 4]
```diff
@@ -12,7 +12,13 @@
 try:
     from typing import get_origin
 except ImportError:
-    from typing_inspect import get_origin
+    def get_origin(typ):
+        result = getattr(typ, "__origin__", None)
+        bases = getattr(result, "__orig_bases__", None)
+        if bases:
+            result = bases[0]
+        return result
```
Comment (dcermak) [inline osc/util/models.py:20]: How about reusing `typing_extensions`?
```suggestion
    from typing_extensions import get_origin
```

File: osc/conf.py [hunk 5]
```diff
@@ -59,382 +64,1071 @@
...
+    # compat function with the config dict
+    def __contains__(self, name):
+        field_name = self._get_field_name(name)
+        return field_name is not None
+
+    # compat function with the config dict
+    def get(self, name, default):
+        try:
+            return self[name]
+        except:
```
[the remark this answers is not in the document]
Comment (dmach) [reply osc/conf.py]: Fixed by throwing and catching KeyError.

[the remark this answers is not in the document]
Comment (dmach) [reply osc/conf.py]: Fixed by throwing and catching KeyError.

File: osc/conf.py (hunk 2 above)
Comment (dmach) [reply osc/conf.py:1096]: Fixed, thanks!

File: osc/util/models.py (hunk 4 above)
Comment (dmach) [reply osc/util/models.py]: It doesn't seem to be available on python 3.6 (tested on opensuse/leap:15.5)
```
Python 3.6.15 (default, Sep 23 2021, 15:41:43) [GCC] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from typing_extensions import get_origin
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'get_origin'
```

File: osc/conf.py (hunk 3 above)
Comment (dmach) [reply osc/conf.py:1134]: Fixed, thanks.

File: osc/connection.py [hunk 6]
```diff
@@ -258,6 +258,10 @@ def http_request(method: str, url: str, headers=None, data=None, file=None, retr
         if purl.scheme == "https":
             ssl_context = oscssl.create_ssl_context()
             ssl_context.load_default_certs()
+
+            if options.cafile or options.capath:
+                ssl_context.load_verify_locations(cafile=options.cafile, capath=options.capath)
```
[the remark this answers is not in the document]
Comment (ziQ) [reply osc/connection.py:263]: ... and with `capath=""`, I get `SSL Error: [X509: INVALID_DIRECTORY] invalid directory (_ssl.c:4154)`

[the remark this answers is not in the document]
Comment (ziQ) [reply osc/connection.py:263]: Nevermind that. With `capath=None` (which you default to) it works fine.

[the remark this answers is not in the document]
Comment (dmach) [reply osc/connection.py:263]: AFAIK `False` is an invalid value. It should either hold a path or be `None`.

[the remark this answers is not in the document]
Comment (ziQ) [reply osc/connection.py:263]: Correct. I hadn't seen how you read the config (and your default values) due to "large diff hidden".

[the remark this answers is not in the document]
Comment (dmach) [reply osc/connection.py:263]: > Correct. I hadn't seen how you read the config (and your default values) due to "large diff hidden".

I've rendered docs for this pull request, that might be helpful too:
https://opensuse-commander.readthedocs.io/en/readthedocs-test/api/osc.conf.html
https://opensuse-commander.readthedocs.io/en/readthedocs-test/oscrc.html

**Comment by birkenfeld** [comment]
One question about this change... so far, I've been using `oscrc` to store config for my own OSC plugin. Is this still supported, and if yes, what is the way to do it? My current way seems too hacky to be proper:
```
class MyOptions(osc.conf.Options):
    new_field: str = Field(...)
osc.conf.Options = MyOptions
```

**Comment by dmach** [comment]
It looks like it worked, but to be honest, I don't know if it was ever supported.
In my opinion plugins should not use oscrc to store their data.
But if there's a regression and there's a reasonable (non-hackish) way of fixing it, I'll do it.

If I broke something or if you need a good support for plugin configs,
could you open a new issue so we don't discuss it in a closed pull request?

**Comment by birkenfeld** [comment]
Sure, thanks for the quick reply!

**Pull request merged**